fortls/parse_fortran.py

```diff
@@ -1124,6 +1212,657 @@ def check_file(self, obj_tree, max_line_length=-1, max_comment_line_length=-1):
+            file_ast.end_scope(ln)
+            log.debug(f'{line} !!! END "{end_scope_word}" scope({ln})')
+            return True
+        return False
+
+    def _parse_do_fixed_format(
+        self,
+        line: str,
+        ln: int,
+        file_ast: fortran_ast,
+        line_label: str,
+        block_id_stack: list[str],
+    ):
+        if (file_ast.current_scope.get_type() == DO_TYPE_ID) and (
+            line_label is not None
+        ):
+            # TODO: try and move to end_scope pattern
+            did_close = False
+            while (len(block_id_stack) > 0) and (line_label == block_id_stack[-1]):
+                file_ast.end_scope(ln)
+                block_id_stack.pop()
+                did_close = True
+                self.parser_debug("DO", self.line, ln, scope=True)
+            if did_close:
+                return True
+        return False
+
+    def _parse_implicit(self, line: str, ln: int, file_ast: fortran_ast):
+        match = FRegex.IMPLICIT.match(line)
+        if match is None:
+            return False
+        err_message = None
+        if file_ast.current_scope is None:
+            err_message = "IMPLICIT statement without enclosing scope"
+        else:
+            if match.group(1).lower() == "none":
+                file_ast.current_scope.set_implicit(False, ln)
+            else:
+                file_ast.current_scope.set_implicit(True, ln)
+        if err_message:
+            file_ast.parse_errors.append(
+                {
+                    "line": ln,
+                    "schar": match.start(1),
+                    "echar": match.end(1),
+                    "mess": err_message,
+                    "sev": 1,
+                }
+            )
+        self.parser_debug("IMPLICIT", self.line, ln)
+        return True
+
+    def _parse_contains(self, line: str, ln: int, file_ast: fortran_ast):
+        match = FRegex.CONTAINS.match(line)
+        if match is None:
+            return False
+        err_message: str = None
+        try:
+            if file_ast.current_scope is None:
+                err_message = "CONTAINS statement without enclosing scope"
+            else:
+                file_ast.current_scope.mark_contains(ln)
+        except ValueError:
+            err_message = "Multiple CONTAINS statements in scope"
+        if err_message:
+            file_ast.parse_errors.append(
+                {
+                    "line": ln,
+                    "schar": match.start(1),
+                    "echar": match.end(1),
+                    "mess": err_message,
+                    "sev": 1,
+                }
+            )
+        self.parser_debug("CONTAINS", self.line, ln)
+        return True
+
+    def _parse_documentation(
+        self,
+        line: str,
+        ln: int,
+        file_ast: fortran_ast,
+        doc_string: str,
+        next_ln_idx: int,
+    ):
+        match = self.COMMENT_LINE_MATCH.match(line)
+        if not match:
```

Comment:
I am convinced that the documentation parsing is wrong. Since it mixes doxygen and FORD styles.

I am not familiar with FORD style but I don't think this pattern is reasonable, where val gets the docstring

``` f08
integer, parameter :: val = 0

!< some doc string
integer, parameter :: foo = 1
```